libbeat/processors/annotate/kubernetes/podwatcher.go

```diff
@@ -0,0 +1,298 @@
+		//Delete old metadata
+		name, endpoints, containers := p.getMetaKeys(oldPod)
+		p.removePodMapEntry(name, endpoints, containers)
+
+		//Process the new pod changes
+		p.enqueue(newPod)
+	}
+
+}
+
+func (p *podWatcher) onPodAddDelete(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	p.enqueue(pod)
+}
+
+func newPodInformer(client kubernetes.Interface, host string, resyncPeriod time.Duration) cache.SharedIndexInformer {
+	sharedIndexInformer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options v1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = fields.Set{api.PodHostField: host}.AsSelector().String()
+				return client.CoreV1().Pods(v1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options v1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = fields.Set{api.PodHostField: host}.AsSelector().String()
+				return client.CoreV1().Pods(v1.NamespaceAll).Watch(options)
+			},
+		},
+		&corev1.Pod{},
+		resyncPeriod,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+	)
+
+	return sharedIndexInformer
+}
+
+func (p *podWatcher) Run() bool {
+	go p.podInformer.Run(p.stop)
+
+	go wait.Until(p.worker, podCheckPeriod, p.stop)
+
+	//Needed to make sure that events don't flow into the annotator before informer is fully set up
+	synced := make(chan struct{})
+	go func() {
+		for {
+			if p.podInformer.HasSynced() == true {
+				close(synced)
+				break
+			}
+		}
+	}()
+	select {
+	case <-time.After(timeout):
+		return false
+	case <-synced:
+		return true
+	}
+
+}
+
+func (p *podWatcher) enqueue(obj interface{}) {
+	// Beware of "xxx deleted" events
+	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+		obj = unknown.Obj
+	}
+	objName, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	p.podQueue.Add(objName)
+}
+
+func (p *podWatcher) worker() {
+	for {
+		func() {
+			key, quit := p.podQueue.Get()
+			if quit {
+				return
+			}
+			defer p.podQueue.Done(key)
+			err := p.syncPod(key.(string))
+			if err != nil {
+				logp.Err("Unable to process %v", key)
+			}
+		}()
+	}
+}
+
+func (p *podWatcher) syncPod(key string) error {
+	p.annotationCache.mu.Lock()
+	defer p.annotationCache.mu.Unlock()
+
+	podObj, ok, err := p.podInformer.GetIndexer().GetByKey(key)
+	switch {
+	case !ok:
+		// pod absence in store means watcher caught the deletion, ensure pod meta info is cleaned
+		logp.Info("Pod has been deleted %v", key)
+	case err != nil:
+		logp.Info("Unable to retrieve pods %v from store: %v", key, err)
+		//p.podQueue.Add(key)
+		return err
+	default:
+		pod, ok := podObj.(*corev1.Pod)
+		if ok {
+			logp.Info("Processing %s:%s", pod.Namespace, pod.Name)
+		}
+
+		name, endpoints, containers := p.getMetaKeys(pod)
+
+		//Pod is in terminating state, remove the metadata
+		if pod.DeletionTimestamp != nil {
+			p.removePodMapEntry(name, endpoints, containers)
```

Comment:
typically, log flow/metric scraping stops after the pod is marked for terminated. we could also move this to another method so that we delete it after the container is completely destroyed.